text/0000-imply-option.md

```diff
@@ -0,0 +1,146 @@
+- Feature Name: imply-option
+- Start Date: 2017-10-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an RFC to reduce common boiler plate code when making use of the `Option` type, providing two functions
+```
+    /// Construct `Some(T)`, conditionally on a boolean.
+    fn Option::<T>::on_pred(bool, T);
+    /// A lazy equivalent of `on_pred`.
+    fn Option::<F: FnOnce() -> T>::lazy_pred(bool, F);
+```
+Similar in intention and motivation to the `Try` trait for `Result`.
+
+# Motivation
+[motivation]: #motivation
+
+This addition will increase the legibility of code segments and assist in defining the thought processes and motivations of programmers through code. The use cases of this addition are problems which are expressable in the following predicate form:
+```
+    P(x) : Predicate on `x`.
+    F(y) : Function on `y`
+    P(x) -> F(y)
+```
+Or the following Rust pseudocode:
+```
+    if P(x) {
+        Some(F(y))
+    } else {
+        None
+    }
+```
+The outcome of this addition will reduce repeated code which introduces bugs during refactoring and present the thought process of the programmer in a clearer fashion through their code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Option` type is useful when your code may or may not yield a return value.
+Such code may looks similar to this:
+```
+    let x = 0;
+    
+    if x == 0 {
+        Some(x)
+    } else {
+        None
+    }
+```
+However only the `if` branch of this code segment is the important part we're concerned about in our code:
+```
+    if x == 0 {
+        Some(x)
+    }
+```
+But the `else` branch is required for returning `None` value if `x == 0` evaluates to false.
+Fortunately Rusts `Option` type has functionality to get rid of the unecessary code:
+```
+    let x = 0;
+    
+    Option::on_pred(x == 0, x)
+```
+This code has the exact same behaviour as our original `if` statement. Our code is however compressed to a single line and our intentions are just as clear.
+Have you spotted the possible issue with this solution introduces however? What about this code:
+```
+    Option::on_pred(false, foo())
+```
+The above line of code will always return `None` and always throw away the result of `foo()` wasting our precious computing power every time our code needs to return `None`.
+Rust has thought ahead of this problem though:
+```
+    Option::lazy_pred(false, foo)
+```
+`Option`s `lazy_pred` function leverages lazy evaluation by taking a function pointer as its second argument. If its first argument evaluates to `true` it will return `Some(foo())` but if its first argument is `false` it returns `None` without having to run `foo`. This solves the problem presented in our earlier example without sacrificing the advantages it gave us.
```

Comment:
I have a feeling with optimizations on Rust would be able to figure this out and not need this method at all.
If you switch it from true to false the code generated completely changes: https://godbolt.org/g/o2WE4h

I could be wrong but this seems like something LLVM will handle fine without us needing the lazy version of this function.

Reply:
This is more about side-effects - it's pretty much`.map(|f| f())`.

Reply:
Or `.map(|_| {...})` on `Option<()>`.

Reply:
Ah okay that makes more sense then.